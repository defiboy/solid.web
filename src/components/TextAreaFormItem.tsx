import React from 'react'

import { Form, Input } from 'antd'
import { FieldProps } from 'formik'

const FormItem = Form.Item
const TextArea = Input.TextArea;

interface InputProps {
  label: string
  placeHolder?: string
}

type InputOwnProps = FieldProps & InputProps

export const TextAreaFormItem: React.FC<InputOwnProps> = ({ label, field, form, placeHolder }: InputOwnProps) => {
  return (
    <FormItem
      htmlFor={field.name}
      label={label}
      hasFeedback={!!form.errors.name}
      validateStatus={form.errors[field.name] && 'error'}
      help={form.errors[field.name]}>
      <TextArea id={field.name} rows={15} data-testid={`textarea-${field.name}`} {...field} placeholder={placeHolder} />
    </FormItem>
  )
}
