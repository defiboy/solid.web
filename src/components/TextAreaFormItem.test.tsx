import React from 'react';
import { Formik, FormikErrors, Field } from 'formik';
import { render, fireEvent } from '@testing-library/react'

import { TextAreaFormItem } from './TextAreaFormItem'

describe('TextAreaFormItem', () => {

    const onSubmitMockHandler = jest.fn()

    interface GenericType {
        name: string,
        url: string
    }

    const initialValues: GenericType = {
        name: '',
        url: ''
    }

    const renderWithWrapperForm = (element: React.ReactElement) => {
        return render(<Formik
            initialValues={initialValues}
            onSubmit={onSubmitMockHandler}
            validate={(items: GenericType) => {
                const errors: FormikErrors<GenericType> = {}
                if (!items.name) {
                    errors.name = 'Required'
                }
                if (!items.url) {
                    errors.url = 'Required'
                }
                return errors
            }}>
            {element}
        </Formik>)
    }

    it('renders an input', async () => {
        const { getByLabelText, getByTestId } = renderWithWrapperForm(
            <Field
                name="url"
                render={(innerProps: any) => (
                    <TextAreaFormItem {...innerProps} label="Blockchain URL" placeHolder="JSON RPC endpoint" />
                )}
            />
        )

        expect(getByLabelText(/Blockchain URL/i)).toBeInTheDocument()
        expect(getByTestId('textarea-url')).toBeInTheDocument()
    })

    it('renders an input with label and placeholder', () => {
        const placeHolder = "JSON RPC endpoint"
        const label = "Blockchain URL"
        const name = "url"

        const { getByTestId } = renderWithWrapperForm(
            <Field
                name={name}
                render={(innerProps: any) => (
                    <TextAreaFormItem {...innerProps} label={label} placeHolder={placeHolder} />
                )}
            />
        )

        const element = getByTestId('textarea-url')

        expect(element).toHaveAttribute('placeHolder', placeHolder)
        expect(element).toHaveAttribute('name', name)
        expect(element).toHaveAttribute('id', name)
        expect(element).toHaveAttribute('class', 'ant-input')
        expect(element).toHaveAttribute('rows', '15')
        expect(element).toHaveValue('')
    })

    it('can be set a new value to the input', () => {
        const placeHolder = "JSON RPC endpoint"
        const label = "Blockchain URL"
        const name = "url"

        const { getByTestId } = renderWithWrapperForm(
            <Field
                name={name}
                render={(innerProps: any) => (
                    <TextAreaFormItem {...innerProps} label={label} placeHolder={placeHolder} />
                )}
            />
        )

        const element = getByTestId('textarea-url')

        expect(element).toHaveValue('')

        const newValue = "newValue"

        fireEvent.change(element, {
            target: {
                value: newValue
            }
        })

        expect(element).toHaveValue(newValue)
    })

    it('should match snapshot', () => {
        const placeHolder = "JSON RPC endpoint"
        const label = "Blockchain URL"
        const name = "url"

        const { container } = renderWithWrapperForm(
            <Field
                name={name}
                render={(innerProps: any) => (
                    <TextAreaFormItem {...innerProps} label={label} placeHolder={placeHolder} />
                )}
            />
        )

        expect(container).toMatchSnapshot()
    })

})
