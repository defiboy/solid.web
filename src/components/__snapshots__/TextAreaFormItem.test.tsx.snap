// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaFormItem should match snapshot 1`] = `
<div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="url"
        title="Blockchain URL"
      >
        Blockchain URL
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control-wrapper"
    >
      <div
        class="ant-form-item-control"
      >
        <span
          class="ant-form-item-children"
        >
          <textarea
            class="ant-input"
            data-testid="textarea-url"
            id="url"
            name="url"
            placeholder="JSON RPC endpoint"
            rows="15"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
